### ISSUE TITLE
Consul leadership transitions with Consul Template and nodes joining/leaving

### ISSUE THREAD

**Issue by darron** [hashicorp/consul-template#331]
A couple weeks ago, @KFishner came out to chat with our team at Datadog and we had mentioned that we had run into a problem with:

1. Consul Template running on every node building a recipe with a number of different services and nodes.
2. nodes joining and leaving once or twice a minute.

We ended up working around that problem by running Consul Template on a single node, stuffing the result into the KV store and using a watch on each node to write the result in place.

That works great, but it's a bit of a hack and wanted to see if there's actually an issue to be fixed. Also - "wait" didn't work properly when we were initially testing but it's in place now wherever we're using Consul Template.

I promised that I'd create a sample environment that duplicated the conditions where this happens and I spent a couple days making this:

https://github.com/darron/consul-bug-report

That's a repo with a Packer recipe and Terraform code to build and spin up a cluster of 120 nodes - instructions on how to duplicate and cause the faults are in the README.

I'm just not sure where to place this - as a Consul issue or a Consul Template issue. I don't think it's completely a Consul Template issue - but not totally sure where it fits best.

I know most of the team is at Gophercon this week - let me know if you've got any questions.

**Comment by slackpad** [comment]
@darron thanks for the awesome repro case! I was out traveling but I'll try to give this a shot by early next week and see what's up.

**Comment by slackpad** [comment]
Just saw this thread as well - linking it for reference - https://groups.google.com/forum/#!topic/consul-tool/NwW9Ygcpejk.

**Comment by darron** [comment]
I'm out travelling after today for 12 days but will try and comment/clarify if needed.

**Comment by slackpad** [comment]
Hi @darron sorry for the delay - I finally had some time to get this going and I was able to stand up the test cluster with the only modification being to the network rules to allow the downloads and all the internal Consul traffic. There was some AWS policy fiddling on my part as well since I was using a new test account.

```diff
diff --git a/consul/security_groups.tf b/consul/security_groups.tf
index a314415..4a01560 100644
--- a/consul/security_groups.tf
+++ b/consul/security_groups.tf
@@ -17,6 +17,20 @@ resource "aws_security_group" "consul" {
         self = true
     }

+    egress {
+        from_port = 0
+        to_port = 65535
+        protocol = "tcp"
+        self = true
+    }
+
+    egress {
+        from_port = 0
+        to_port = 65535
+        protocol = "udp"
+        self = true
+    }
+
     // These are for maintenance
     ingress {
         from_port = 22
@@ -24,4 +38,18 @@ resource "aws_security_group" "consul" {
         protocol = "tcp"
         cidr_blocks = ["0.0.0.0/0"]
     }
+
+    egress {
+       from_port = 80
+        to_port = 80
+        protocol = "tcp"
+        cidr_blocks = ["0.0.0.0/0"]
+    }
+
+    egress {
+       from_port = 443
+        to_port = 443
+        protocol = "tcp"
+        cidr_blocks = ["0.0.0.0/0"]
+    }
 }
```

Now that we've got a matching test case we will be able to try the upcoming 0.6 performance improvements on here to see how things change. Will keep you posted.

**Comment by darron** [comment]
Thanks!

**Comment by sethvargo** [comment]
Hi @darron 

As per your conversations with @slackpad, this will scale much better in Consul 0.6. Do you want to keep this open until Consul 0.6 is released?

**Comment by darron** [comment]
It's OK if you close it - just glad 0.6 and the read improvements will hopefully make this a non-issue.

Take care - much thanks to @slackpad for looking at this with me.

**Comment by slackpad** [comment]
Hi @darron - spun up the test cluster with 120 nodes tonight. With Consul 0.5.2 I was getting leader elections after bouncing even one client. After upgrading to a Consul 0.6.0 release candidate build with the new state store I wasn't able to get the cluster to flap, even which cycling 3 at one time. We got some pretty high query volumes as well.

![image](https://cloud.githubusercontent.com/assets/673509/10780450/bf33abb2-7cfa-11e5-874d-eef1cb6d91ce.png)

I hope to have a build people can experiment with tomorrow. Thanks!

**Comment by slackpad** [comment]
Let things run overnight (all with 0.5.2 clients) and saw one organic leader election and one leader election when I cycled three clients rapidly this morning. Good news is GC performance looks flat through all of this (the spike here is right before I upgraded the servers):

![image](https://cloud.githubusercontent.com/assets/673509/10793570/b16490ba-7d4e-11e5-80b0-853bf4eca0d7.png)

I have a theory that the old clients are bombarding the server with TCP connections and causing issues (especially with these small, single-core instances), so I'm going to re-roll the test cluster with 0.6.0 everywhere and see if things look more stable.

**Comment by darron** [comment]
Very cool. Looking forward to this next release @slackpad - thanks!

**Comment by slackpad** [comment]
Just wanted to update with a little more testing results.

After I upgraded all the node agents to the new release, everything fell over hard! Chased that down and fixed it here:

https://github.com/hashicorp/consul/pull/1357

With that in place things look pretty stable. I automated 3 nodes randomly restarting Consul and went for long periods without any leader elections. It's interesting that if there's any network rattiness then it'll start hitting consul-template way more in this test configuration as serf health checks flap, amplifying everything. We managed to survive excursions into 150k RPC calls / 2 min with only 2 leader elections during the test period.

![image](https://cloud.githubusercontent.com/assets/673509/10807563/4b9d5820-7d9e-11e5-87ae-d3aaf501a6e8.png)

This test setup has been super useful! Just wanted to thank you @darron again for taking the time to get this going.
